vochain/account.go

```diff
@@ -461,250 +473,253 @@ func (v *State) SetAccount(accountAddress common.Address, account *Account) erro
+		return fmt.Errorf("burnTxCostIncrementNonce: %w", err)
+	}
 	// get account
 	acc, err := v.GetAccount(accountAddress, false)
 	if err != nil {
-		return fmt.Errorf("subtractCostIncrementNonce: %w", err)
+		return fmt.Errorf("burnTxCostIncrementNonce: %w", err)
 	}
 	if acc == nil {
 		return ErrAccountNotExist
 	}
-	// get tx cost
-	cost, err := v.TxCost(txType, false)
-	if err != nil {
-		return fmt.Errorf("subtractCostIncrementNonce: %w", err)
-	}
-	// increment nonce
-	if txType != models.TxType_COLLECT_FAUCET {
-		acc.Nonce++
-	}
 	if cost != 0 {
 		// send cost to burn address
 		burnAcc, err := v.GetAccount(BurnAddress, false)
 		if err != nil {
-			return fmt.Errorf("subtractCostIncrementNonce: %w", err)
+			return fmt.Errorf("burnTxCostIncrementNonce: %w", err)
 		}
 		if burnAcc == nil {
-			return fmt.Errorf("subtractCostIncrementNonce: burn account does not exist")
+			return fmt.Errorf("burnTxCostIncrementNonce: burn account does not exist")
 		}
 		if err := acc.Transfer(burnAcc, cost); err != nil {
-			return fmt.Errorf("subtractCostIncrementNonce: %w", err)
+			return fmt.Errorf("burnTxCostIncrementNonce: %w", err)
 		}
 		log.Debugf("burning fee for tx %s with cost %d from account %s", txType.String(), cost, accountAddress.String())
 		if err := v.SetAccount(BurnAddress, burnAcc); err != nil {
-			return fmt.Errorf("subtractCostIncrementNonce: %w", err)
+			return fmt.Errorf("burnTxCostIncrementNonce: %w", err)
 		}
 	}
+	acc.Nonce++
 	if err := v.SetAccount(accountAddress, acc); err != nil {
-		return fmt.Errorf("subtractCostIncrementNonce: %w", err)
+		return fmt.Errorf("burnTxCostIncrementNonce: %w", err)
 	}
 	return nil
 }
 
 // MintTokensTxCheck checks if a given MintTokensTx and its data are valid
-func MintTokensTxCheck(vtx *models.Tx, txBytes, signature []byte, state *State) (common.Address, uint64, error) {
+func MintTokensTxCheck(vtx *models.Tx, txBytes, signature []byte, state *State) error {
+	if vtx == nil || txBytes == nil || signature == nil || state == nil {
+		return fmt.Errorf("invalid parameters")
+	}
 	tx := vtx.GetMintTokens()
-	// check signature available
-	if signature == nil || tx == nil || txBytes == nil {
-		return common.Address{}, 0, fmt.Errorf("missing signature and/or transaction")
+	if tx == nil {
+		return fmt.Errorf("invalid tx")
 	}
-	// check value
 	if tx.Value <= 0 {
-		return common.Address{}, 0, fmt.Errorf("invalid value")
+		return fmt.Errorf("invalid value")
 	}
-	// check to
-	if len(tx.To) != types.EntityIDsize || bytes.Equal(tx.To, types.EthereumZeroAddress[:]) {
-		return common.Address{}, 0, fmt.Errorf("invalid To address")
-	}
-	// get treasurer
-	treasurer, err := state.Treasurer(false)
-	if err != nil {
-		return common.Address{}, 0, err
-	}
-	// check nonce
-	if tx.Nonce != treasurer.Nonce {
-		return common.Address{}, 0, fmt.Errorf("invalid nonce %d, expected: %d", tx.Nonce, treasurer.Nonce)
+	if len(tx.To) == 0 || bytes.Equal(tx.To, types.EthereumZeroAddress.Bytes()) {
```

Comment:
What the zero address did to you?? :)

Reply:
nothing haha, is a basic check all the eth code does to prevent operating with the zero account, and coming from that world i took this habit.
Anyways, deleted all zeroaddr checks.

Reply:
update: still checking for delegates operation as the tx.Account is a field that can be different than txSender for those cases.